parser-typechecker/src/Unison/Codebase/Causal.hs

```diff
@@ -112,6 +113,24 @@ data Tails h
 
 type Deserialize m h e = RawHash h -> m (Raw h e)
 
+cachedRead :: Monad m
+           => Cache.Cache m (RawHash h) (Causal m h e)
+           -> Deserialize m h e
+           -> RawHash h -> m (Causal m h e)
+cachedRead cache deserializeRaw h = Cache.lookup cache h >>= \case
+  Nothing -> do
+    raw <- deserializeRaw h
+    causal <- pure $ case raw of
+      RawOne e              -> One h e
+      RawCons e tailHash    -> Cons h e (tailHash, read tailHash)
+      RawMerge e tailHashes -> Merge h e $
+        Map.fromList [(h, read h) | h <- toList tailHashes ]
+    Cache.insert cache h causal
+    pure causal
+  Just causal -> pure causal
+  where
+    read = cachedRead cache deserializeRaw
```

Comment:
Here's where the cache gets used, very similar to existing `Causal.read`, but with some calls to `Cache.insert` and `Cache.lookup`.